File: src/main/java/it/giaquinto/springberry/controller/SpringBerryController.java
package it.giaquinto.springberry.controller;

import it.giaquinto.springberry.model.http.HttpRequest;
import it.giaquinto.springberry.component.MathComponent;
import it.giaquinto.springberry.component.StringComponent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;


public abstract class SpringBerryController {

    private StringComponent stringComponent;

    private MathComponent mathComponent;

    private final Logger logger = LoggerFactory.getLogger(this.getClass());

    public final StringComponent getStringUtils() {
        return stringComponent;
    }

    @Autowired
    public final void setStringUtils(final StringComponent stringComponent) {
        this.stringComponent = stringComponent;
    }

    public final MathComponent getMathUtils() {
        return mathComponent;
    }

    @Autowired
    public final void setMathUtils(final MathComponent mathComponent) {
        this.mathComponent = mathComponent;
    }

    public final Logger getLogger() {
        return logger;
    }

    public abstract String ID();

    /**
     * Specific the accepted Http Method for the controller
     * @return HttpRequest[]
     */
    public abstract HttpRequest[] controllerAcceptedRequest();
}

File: src/main/java/it/giaquinto/springberry/spring/controller/SpringBerryController.kt
package it.giaquinto.springberry.spring.controller

import it.giaquinto.springberry.common.logger.SpringBerryLoggerBean
import it.giaquinto.springberry.spring.Identifier
import it.giaquinto.springberry.spring.Rest
import it.giaquinto.springberry.spring.RestRadix
import it.giaquinto.springberry.spring.model.SpringBerryModel
import it.giaquinto.springberry.spring.model.api.ApiResult
import it.giaquinto.springberry.spring.model.http.HttpRequest
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import org.springframework.beans.factory.annotation.Autowired

abstract class SpringBerryController<T> {

    abstract val identifier: Identifier

    abstract val uniqueRestRadix: RestRadix



    abstract fun defaultResponse(): ApiResult<T>

    protected open fun controllerAcceptedMethod(): ApiResult<Array<HttpRequest?>?> =
        ApiResult.Success(_data = arrayOf(HttpRequest.GET))

    companion object {
        const val radixControllerAcceptedRequest: Rest = "httpAcceptedMethod"
    }

}